### PR TITLE
Integrate DRM into the react-native-video player - Android

### PR DESCRIPTION
**Pull request by marcinnowak2**

### CODE REVIEW

File: examples/basic/android/settings.gradle [hunk 1]
```diff
@@ -8,5 +8,9 @@ include ':app',
 //project(':react-native-video').projectDir = new File(rootProject.projectDir, '../../android')
 //project(':react-native-video-exoplayer').projectDir = new File(rootProject.projectDir, '../../android-exoplayer')
 
-project(':react-native-video').projectDir = new File(rootProject.projectDir, '../../android')
-project(':react-native-video-exoplayer').projectDir = new File(rootProject.projectDir, '../../android-exoplayer')
+project(':react-native-video').projectDir = new File(rootProject.projectDir, '../../../android')
+project(':react-native-video-exoplayer').projectDir = new File(rootProject.projectDir, '../../../android-exoplayer')
+
+//FIXME:
+include ':fea_dice_shield_dlm'
```
Comment (marcinnowak2) [inline examples/basic/android/settings.gradle]: Do we know how we would like to pull dice-shield lib?

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java [hunk 2]
```diff
@@ -279,16 +319,37 @@ private void initializePlayer() {
...
-                        new DefaultDashChunkSource.Factory(mediaDataSourceFactory), mainHandler, null);
+                return new DashMediaSource.Factory(
+                        new DefaultDashChunkSource.Factory(mediaDataSourceFactory),
+                        buildDataSourceFactory(false))
+                        .setManifestParser(
+                                new FilteringManifestParser<>(
+                                        new DashParser(new DashParser.Callback() {
+                                            @Override
+                                            public void onManifestParsed(com.google.android.exoplayer2.source.dash.manifest.DashManifest manifest) {
+                                                Log.d("Player", "onManifestParsed() manifest=" + manifest);
```
Comment (marcinnowak2) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java:486]: Shall I comment out or remove those logs completely?

File: examples/basic/android/settings.gradle [hunk 3]
```diff
@@ -8,5 +8,9 @@ include ':app',
 //project(':react-native-video').projectDir = new File(rootProject.projectDir, '../../android')
```
Comment (vdmtrv) [inline examples/basic/android/settings.gradle]: Is this commented out code still needed?

Comment (marcinnowak2) [reply examples/basic/android/settings.gradle]: No. I guess I can remove it.

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java [hunk 4]
```diff
@@ -279,16 +319,37 @@ private void initializePlayer() {
         }
     }
 
-    private MediaSource buildMediaSource(Uri uri, String overrideExtension) {
-        int type = Util.inferContentType(!TextUtils.isEmpty(overrideExtension) ? "." + overrideExtension
-                : uri.getLastPathSegment());
+    @SuppressWarnings("unchecked")
+    private MediaSource buildMediaSource(Uri uri, @Nullable String overrideExtension) {
```
Comment (basadi-dev) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java:470]: Would be good to keep it consistent, use `@NonNull` or `@Nullable` annotations

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java [hunk 5]
```diff
@@ -300,6 +361,84 @@ private MediaSource buildMediaSource(Uri uri, String overrideExtension) {
...
+    private void releaseMediaDrm() {
+        if (mediaDrm != null) {
+            mediaDrm.release();
+            mediaDrm = null;
+        }
+        drmCallback = null;
+    }
+
+    private List<?> getOfflineStreamKeys(Uri uri) {
+        return Collections.emptyList();
```
Comment (basadi-dev) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: not using the uri.

Could use `@Nullable`/`@NonNull` annotations

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java [hunk 6]
```diff
@@ -279,16 +319,37 @@ private void initializePlayer() {
...
+
+                                                if (manifest instanceof DashManifest && actionToken != null) {
+                                                    List kids = ((DashManifest) manifest).getDefaultKIds();
+                                                    Log.d("Player", "onManifestParsed() KIds=" + kids);
+                                                    String header = Utils.createXDrmInfoHeader(Utils.getSystem(actionToken.getDrmScheme()), kids);
+                                                    Log.d("Player", "onManifestParsed() XDrmHeader=" + header);
+                                                    drmCallback.setKeyRequestProperty("X-DRM-INFO", header);
+                                                }
+                                            }
+                                        }), (List<RepresentationKey>) getOfflineStreamKeys(uri)))
```
Comment (basadi-dev) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: Is the casting requried?

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java [hunk 7]
```diff
@@ -300,6 +361,84 @@ private MediaSource buildMediaSource(Uri uri, String overrideExtension) {
         }
     }
 
+    private DefaultDrmSessionManager createDrmSessionManager(ActionToken drmParam) throws UnsupportedDrmException {
```
Comment (basadi-dev) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: Could be marked as `@Nullable`

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java [hunk 8]
```diff
@@ -300,6 +361,84 @@ private MediaSource buildMediaSource(Uri uri, String overrideExtension) {
...
+                errorStringId = com.dice.shield.downloads.R.string.error_drm_unknown;
+                break;
+        }
+
+        final String errorString = getContext().getString(errorStringId);
+        eventEmitter.error(errorString, exception);
+    }
+
+    private DefaultDrmSessionManager<FrameworkMediaCrypto> buildDrmSessionManagerV18(
+            UUID uuid, String licenseUrl, HashMap<String, String> requestProperties, boolean multiSession)
```
Comment (basadi-dev) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: Could it be a `Map<String, String>` so it's not a specifc implemntation of Maps?

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java [hunk 9]
```diff
@@ -300,6 +361,84 @@ private MediaSource buildMediaSource(Uri uri, String overrideExtension) {
...
+            throws UnsupportedDrmException {
+        HttpDataSource.Factory licenseDataSourceFactory =
+                (DlmWrapper.getInstance(getContext().getApplicationContext())).buildHttpDataSourceFactory(/* listener= */ null);
```
Comment (basadi-dev) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: Could NullPointerException be thrown?

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java [hunk 10]
```diff
@@ -300,6 +361,84 @@ private MediaSource buildMediaSource(Uri uri, String overrideExtension) {
...
+                drmCallback.setKeyRequestProperty(key, requestProperties.get(key));
+            }
+        }
+        releaseMediaDrm();
+        mediaDrm = FrameworkMediaDrm.newInstance(uuid);
+        this.drmCallback = drmCallback;
+        return new DefaultDrmSessionManager<>(uuid, mediaDrm, drmCallback, null, multiSession);
+    }
+
+    private void releaseMediaDrm() {
```
Comment (basadi-dev) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java:573]: Do not think we have multi-threaded operations running here, but it feels like that release should be a synchronize method.

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java [hunk 11]
```diff
@@ -697,22 +840,29 @@ public void onMetadata(Metadata metadata) {
...
+
+            initializePlayer();
 
             if (!isOriginalSourceNull && !isSourceEqual) {
                 reloadSource();
             }
         }
     }
 
+    public void setSrc(final Uri uri, final String extension, Map<String, String> headers) {
```
Comment (basadi-dev) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: Could use annotations.

**Review by basadi-dev — APPROVED** [review]
General minor comments and questions.

File: examples/basic/android/settings.gradle (hunk 3 above)
Comment (marcinnowak2) [reply examples/basic/android/settings.gradle]: Removed.

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java (hunk 6 above)
Comment (marcinnowak2) [reply android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: No. I have changed the method signature so there is not casting now.

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java (hunk 7 above)
Comment (marcinnowak2) [reply android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: I have made it nonnull as it is not possible to create drm manager without drm params.

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java (hunk 8 above)
Comment (marcinnowak2) [reply android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: Done.

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java (hunk 9 above)
Comment (marcinnowak2) [reply android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: No. DlmWrapper is a singleton.

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java (hunk 10 above)
Comment (marcinnowak2) [reply android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java:573]: It should be done on one thread only so I don't think it is needed. However I can see it is not called when player is released and I guess it should be.

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java (hunk 5 above)
Comment (marcinnowak2) [reply android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: Changed method signature and added annotations.

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java (hunk 11 above)
Comment (marcinnowak2) [reply android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: Yes. Done.

File: android-exoplayer/build.gradle [hunk 12]
```diff
@@ -25,12 +25,13 @@ dependencies {
...
     // All support libs must use the same version
     compile "com.android.support:support-annotations:${safeExtGet('supportLibVersion', '+')}"
-    compile "com.android.support:support-compat:${safeExtGet('supportLibVersion', '+')}"
+    compile "com.android.support:appcompat-v7:${safeExtGet('supportLibVersion', '+')}"
     compile "com.android.support:support-media-compat:${safeExtGet('supportLibVersion', '+')}"
 
     compile('com.google.android.exoplayer:extension-okhttp:2.8.2') {
         exclude group: 'com.squareup.okhttp3', module: 'okhttp'
     }
     compile 'com.squareup.okhttp3:okhttp:3.11.0'
```
Comment (basadi-dev) [inline android-exoplayer/build.gradle:34]: Could we make the version configurable as well (similar to support:compat libs)

**Pull request merged**